klippa/src/hmtx.rs

```diff
@@ -0,0 +1,79 @@
+//! impl subset() for hmtx
+
+use crate::{Plan, Subset, SubsetError};
+use read_fonts::TopLevelTable;
+use std::cmp;
+use write_fonts::tables::{hmtx::Hmtx, hmtx::LongMetric};
+
+impl Subset for Hmtx {
+    fn subset(&mut self, plan: &mut Plan) -> Result<bool, SubsetError> {
+        let gids = &plan.glyph_ids;
+        if gids.is_empty() {
+            Err(SubsetError::SubsetTableError(Hmtx::TAG))
+        } else {
+            let num_glyphs = self.h_metrics.len() + self.left_side_bearings.len();
+            let mut num_long_metrics = cmp::min(num_glyphs, 0xFFFF_usize);
+            let last_advance = get_gid_advance(&self.h_metrics, num_glyphs - 1);
+            for gid in gids.iter().rev() {
+                let glyph_id = gid.to_u16() as usize;
+                // skip invalid gids
+                if glyph_id >= num_glyphs {
+                    continue;
+                }
+
+                let advance = get_gid_advance(&self.h_metrics, glyph_id);
+                if num_long_metrics == 1 || last_advance != advance {
+                    break;
+                }
+                num_long_metrics -= 1;
+            }
+
+            let mut new_metrics = Vec::<LongMetric>::with_capacity(num_long_metrics);
+            let mut new_side_bearings = Vec::<i16>::new();
+            for gid in gids {
+                let glyph_id = gid.to_u16() as usize;
+                // skip invalid gids
+                if glyph_id >= num_glyphs {
+                    break;
+                }
+
+                let side_bearing =
+                    get_gid_side_bearing(&self.h_metrics, &self.left_side_bearings, glyph_id);
+                if glyph_id < num_long_metrics {
+                    let advance = get_gid_advance(&self.h_metrics, glyph_id);
+                    new_metrics.push(LongMetric {
+                        advance,
+                        side_bearing,
+                    });
+                } else {
+                    new_side_bearings.push(side_bearing);
+                }
+            }
+
+            // update num_of_hmetrics in the plan so we can update hhea table
+            plan.num_h_metrics = new_metrics.len() as u16;
+            // temporary: update num_output_glyphs in the plan for maxp table, will do this during planning phase
+            plan.num_output_glyphs = new_metrics.len() as u16 + new_side_bearings.len() as u16;
+            self.h_metrics = new_metrics;
+            self.left_side_bearings = new_side_bearings;
+
+            Ok(true)
+        }
+    }
+}
+
+fn get_gid_advance(metrics: &[LongMetric], gid: usize) -> u16 {
+    if gid >= metrics.len() {
+        metrics.last().unwrap().advance
+    } else {
+        metrics[gid].advance
+    }
+}
+
+fn get_gid_side_bearing(metrics: &[LongMetric], side_bearings: &[i16], gid: usize) -> i16 {
+    if gid < metrics.len() {
```

Comment:
use `metrics.get(gid)`, it performs the bounds check for you and returns an `Option`.